Fix build errors when targeting 14.5 with the latest SDK

tadeuzagallo opened the pull request:
#### e6cbba03b97bd8a9690c90630b4a22518cd9a3c9
<pre>
Fix build errors when targeting 14.5 with the latest SDK
<a href="https://bugs.webkit.org/show_bug.cgi?id=287703">https://bugs.webkit.org/show_bug.cgi?id=287703</a>
<a href="https://rdar.apple.com/144867680">rdar://144867680</a>

Reviewed by Elliott Williams.

Add conditional checks around APIs that are not available when targetting 14.5

* Configurations/CommonBase.xcconfig:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm:
(WebKit::RemoteAcceleratedEffectStack::initEffectsFromMainThread):
(WebKit::RemoteAcceleratedEffectStack::applyEffectsFromScrollingThread const):
(WebKit::RemoteAcceleratedEffectStack::clear):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::useNetworkLoader):
* Source/WebKit/WebKitSwift/TextExtraction/WKWebView+TextExtraction.swift:
(WKWebView._intelligenceCollectRemoteContent(in:remoteContextWrapper:)):
* Source/WebKit/WebKitSwift/WritingTools/PlatformIntelligenceTextEffectView.swift:

Canonical link: <a href="https://commits.webkit.org/293521@main">https://commits.webkit.org/293521@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3ad0cb83952c8c19a4f74e9d217ce60954c7199

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99016 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18654 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8900 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104145 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49609 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101061 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18948 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27103 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75377 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32503 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102020 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14408 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89425 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55828 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14200 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7399 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48985 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/91706 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84150 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7475 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106511 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/97646 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26113 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19132 "Found 1 new test failure: imported/w3c/web-platform-tests/wasm/webapi/esm-integration/script-src-blocks-wasm.tentative.sub.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84337 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26488 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85621 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83840 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21295 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28506 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6177 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19850 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26066 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31252 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/121262 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25886 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33906 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29206 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27460 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Configurations/CommonBase.xcconfig:
```diff
@@ -97,17 +97,17 @@ CLANG_CXX_STANDARD_LIBRARY_HARDENING = extensive;
 WK_LIBCPP_ASSERTIONS_CFLAGS = $(WK_LIBCPP_ASSERTIONS_CFLAGS_$(WK_PLATFORM_NAME));
-WK_LIBCPP_ASSERTIONS_CFLAGS_iphoneos = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_IOS_17)$(WK_IOS_18));
-WK_LIBCPP_ASSERTIONS_CFLAGS_iphonesimulator = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_IOS_17)$(WK_IOS_18));
+WK_LIBCPP_ASSERTIONS_CFLAGS_iphoneos[sdk=iphone*17*] = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_IOS_17)$(WK_IOS_18));
```
For example, this is already checking for target version, is that working incorrectly somehow? I don't think that it will be easy to understand why we use both `[sdk=iphone*17*]` and `WK_IOS_17` checks on the same line.

Besides, we don't support iOS 17 at all, neither as target, nor as SDK.

emw-apple replied:
My guess is that the libcpp assertions feature is entirely based on SDK version, not deployment target, because its purpose is to toggle some behavior in C++ stdlib headers. So the right thing to do is remove all the target checks (WK_IOS_17, WK_MACOS_1400, etc.) and implement this using `[sdk=...]` conditions only.

tadeuzagallo replied:
I just applied the same logic from Mac, which was what I was testing, so maybe I should restrict this change to Mac.

My understanding, and it seems to have worked locally, is that `WK_IOS_17` is testing the target, while `[sdk=iphone*17*]` is testing the SDK version being used to compile. Is that not correct?

aproskuryakov replied:
That's likely correct (I didn't check). Are you saying that libcpp assertions require both new SDK and new runtime?

emw-apple commented on Configurations/CommonBase.xcconfig:
```diff
@@ -97,17 +97,17 @@ CLANG_CXX_STANDARD_LIBRARY_HARDENING = extensive;
+WK_LIBCPP_ASSERTIONS_CFLAGS_iphonesimulator[sdk=iphone*17*] = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_IOS_17)$(WK_IOS_18));
 WK_LIBCPP_ASSERTIONS_CFLAGS_IOS_SINCE_17_IOS_BEFORE_18 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
-WK_LIBCPP_ASSERTIONS_CFLAGS_maccatalyst = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_MACCATALYST_14)$(WK_MACCATALYST_15));
+WK_LIBCPP_ASSERTIONS_CFLAGS_maccatalyst[sdk=mac*14*] = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_MACCATALYST_14)$(WK_MACCATALYST_15));
```
nit: `mac*14` is unnecessary
```suggestion
WK_LIBCPP_ASSERTIONS_CFLAGS_maccatalyst[sdk=macosx14*] = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_MACCATALYST_14)$(WK_MACCATALYST_15));
```

aproskuryakov commented on Source/WebKit/WebKitSwift/WritingTools/PlatformIntelligenceTextEffectView.swift:
```diff
@@ -484,7 +484,9 @@ struct PlatformIntelligenceTextEffectID: Hashable {
         UIView.animate(animation, changes: changes)
 #else
-        NSAnimationContext.animate(animation, changes: changes)
+        if #available(macOS 15, iOS 18, tvOS 18, watchOS 10, *) {
```
This is a runtime check, which we don't need in WebKit, as we are not striving to have a single binary that runs everywhere. `#if @available` is compile time. Furthermore,
- no need for embedded platform checks, because this is the branch where we can't import UIKit;
- no need for embedded platform checks, as these are the oldest versions supported.

As mentioned in another comment, it would be ideal to centralize HAVE-style definitions in one file, so that we have a meaningful name for the checks, and reduce the risk of using slightly different checks in different parts of the code.

tadeuzagallo replied:
I spent some time on this, and I could not figure out how make it into a compiler check. Unfortunately, it doesn't look like `#if @available` is valid syntax, and all the valid expressions after `#if` according to the grammar don't handle this operation. My guess is that either I'm missing something, or they do it by design.

Using a definition as you suggested is probably the best way to go, I'll look into that, just didn't change it yet as I don't know this code well enough to understand everything we should compile conditionally. That will be the next step.

emw-apple commented on Source/WebKit/WebKitSwift/TextExtraction/WKWebView+TextExtraction.swift:
```diff
@@ -81,14 +81,16 @@ extension WKWebView {
-            self._requestTextExtraction(in: visibleRect) { item in
-                if let item {
-                    collector.collect(createIntelligenceElement(item: item))
+        if #available(macOS 15, *) {
```
If I'm reading this right, you're saying that this function does a runtime check for macOS 15, and if the system does not pass the check it silently does _nothing_. What we probably want instead is compile-time `@available` attributes on the extension or this particular function. Have you tried that?

tadeuzagallo replied:
Yes, that was my first approach too, but unfortunately it doesn't compile since it's an `override` declaration, so the compiler errors if the availability doesn't match the original/base declaration.

emw-apple replied:
Well, the `#if canImport(UIIntelligenceSupport)` guard is currently being used to only compile this on iOS 18 / macOS Sequoia and above. I'm guessing the ObjC declaration this overrides works the same way, since this code has to do with text extraction features that shipped in those releases.

IMO this change must match the status quo, and therefore it _must not_ compile this on downlevels, as the underlying framework that powers this class it is not present. Issues like the one here are a sign that we're doing the wrong thing, and backporting new features to older releases.

I think this also highlights a risk of shifting to a back-deployed build: during the switch, we risk enabling code in downlevels that was previous compiled out. This may confuse other logic in WebKit using respondsToSelector or type existence checks to determine whether text extraction is available. It's important to get this right; we don't want to accidentally enable broken features as part of the back-deployed build.

emw-apple reviewed (changes requested):
The three `@available` attributes on so many declarations is quite unfortunate. I think that we really need to come up with a way to have <wtf/Platform.h> equivalent compilation conditions in Swift, so that we can write
`#if ENABLE_FOO ...` instead. Sprinkling `@available` throughout the codebase won't scale, and as Alexey notes, part of how you've worked around the lack of platform-flag-like semantics is with runtime checks, which don't really make sense for WebKit.

I think we should chat about whether we want to land this at all in its current state, since we don't have imminent plans to actually start building this way. I'm not sure whether we should check this in now to vs. wait until we have a more forward-looking plan and then clean it up.

emw-apple commented on Configurations/CommonBase.xcconfig:
```diff
@@ -99,18 +99,9 @@ CLANG_CXX_STANDARD_LIBRARY_HARDENING = extensive;
-WK_LIBCPP_ASSERTIONS_CFLAGS_watchos = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_WATCHOS_10)$(WK_WATCHOS_11));
-WK_LIBCPP_ASSERTIONS_CFLAGS_watchsimulator = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_WATCHOS_10)$(WK_WATCHOS_11));
-WK_LIBCPP_ASSERTIONS_CFLAGS_WATCHOS_SINCE_10_WATCHOS_BEFORE_11 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
+WK_LIBCPP_ASSERTIONS_CFLAGS = ;
```
nit: Remove this line, empty default value is implicit.

emw-apple commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1818,6 +1818,16 @@
 #endif
 
+#if !defined(HAVE_WRITING_TOOLS) \
+    && __has_include(<WritingTools/WritingTools.h>) \
```
Do we really need this has_include check? WritingTools.framework should be in all the SDKs we build with since it shipped last year.
```suggestion
```

tadeuzagallo replied:
I will double check, I'm not 100% sure now, but I believe it was failing to compile in EWS without it

tadeuzagallo replied:
Indeed, if you look at https://github.com/WebKit/WebKit/commit/1ce3567b28b06f9a26a94e0bffad3c9940b71c39 I started without it, then you can see all the bots failed in https://github.com/WebKit/WebKit/pull/40625#issuecomment-2715664378

emw-apple replied:
I think what's happening is that WritingTools is a private framework, and someone will need to add more to `WritingToolsSPI.h` to get this building publicly. We can't write USE_APPLE_INTERNAL_SDK here to make this more obvious, so probably fine to leave for now.

emw-apple commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1904,3 +1914,13 @@
 #define HAVE_RE_STEREO_CONTENT_SUPPORT 0
 #endif
+
+#if !defined(HAVE_NWSETTINGS_UNIFIED_HTTP) && defined(NW_SETTINGS_HAS_UNIFIED_HTTP) && (!PLATFORM(MAC) || __MAC_OS_X_VERSION_MIN_REQUIRED >= 150000)
```
NW_SETTINGS_HAS_UNIFIED_HTTP comes from Network.framework, so I'm pretty sure it won't work here, and `defined(NW_SETTINGS_HAS_UNIFIED_HTTP)` will always be false.

I would recommend seeing if we _really_ need to check that header, or if we can just use *_MIN_REQUIRED versions to figure out whether to enable this feature or not.

emw-apple commented on Source/WebCore/dom/DocumentFullscreen.h:
```diff
@@ -65,7 +65,7 @@ class DocumentFullscreen final : public CanMakeWeakPtr<DocumentFullscreen> {
     Ref<Document> protectedDocument() const { return m_document.get(); }
     Page* page() const { return document().page(); }
-    LocalFrame* frame() const { return document().frame(); }
+    LocalFrame* frame() const;
```
Why the change here?

tadeuzagallo replied:
It failed to compile due to the `frame()` function being implemented in an inline header. Not sure why it came up now though to be honest

emw-apple commented on Source/WebKit/UIProcess/API/Swift/DownloadCoordinator.swift:
```diff
@@ -28,6 +28,9 @@ import Foundation
 @_spi(Private)
+@available(WK_IOS_TBA, WK_MAC_TBA, WK_XROS_TBA, *)
+@available(watchOS, unavailable)
+@available(tvOS, unavailable)
```
It's interesting that we have to add these at all, since this code is already guarded by `ENABLE_SWIFTUI`, which is based on the WK_ENABLE_SWIFTUI build setting that you've set based on deployment target. Still, it seems right to add availability annotations, since these are SPI.

cc @rr-codes

emw-apple commented on Source/WebKit/WebKitSwift/TextExtraction/WKWebView+TextExtraction.swift:
```diff
@@ -21,7 +21,7 @@
 // THE POSSIBILITY OF SUCH DAMAGE.
 
-#if canImport(UIIntelligenceSupport)
+#if HAVE_UIINTELLIGENCESUPPORT_FRAMEWORK && canImport(UIIntelligenceSupport)
```
Seems safe to remove the canImport? 
```suggestion
#if HAVE_UIINTELLIGENCESUPPORT_FRAMEWORK
```

tadeuzagallo replied:
This one was for sure causing a build error in EWS, I started without it then had to add it back unfortunately. I considered moving the check to the PlatformHave flag, but couldn't find an equivalent

emw-apple replied:
I think it's a similar situation to HAVE_WRITING_TOOLS, where the framework is private and we don't have an SPI module available for public use. I think it would make more sense to do what HAVE_WRITING_TOOLS does, and add a `__has_include(<UIIntelligenceSupport/UIIntelligenceSupport.h>)` clause here: https://github.com/WebKit/WebKit/blob/12e840378db6d373821660b7317a98170dad399e/Source/WTF/wtf/PlatformHave.h#L1790-L1797

emw-apple reviewed (changes requested):
Looks good, glad to see this coming together! I think you have a bug with NW_SETTINGS_HAS_UNIFIED_HTTP, otherwise I just have a some small questions and suggestions:

mwyrzykowski reviewed (commented):
Looks good but I defer to @emw-apple for approval

emw-apple reviewed (approved):
r+, with a small refactoring request. Thank you for iterating on this!

tadeuzagallo committed:
Fix build errors when targeting 14.5 with the latest SDK
https://bugs.webkit.org/show_bug.cgi?id=287703
rdar://144867680

Reviewed by Elliott Williams.

Add conditional checks around APIs that are not available when targetting 14.5

* Configurations/CommonBase.xcconfig:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm:
(WebKit::RemoteAcceleratedEffectStack::initEffectsFromMainThread):
(WebKit::RemoteAcceleratedEffectStack::applyEffectsFromScrollingThread const):
(WebKit::RemoteAcceleratedEffectStack::clear):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::useNetworkLoader):
* Source/WebKit/WebKitSwift/TextExtraction/WKWebView+TextExtraction.swift:
(WKWebView._intelligenceCollectRemoteContent(in:remoteContextWrapper:)):
* Source/WebKit/WebKitSwift/WritingTools/PlatformIntelligenceTextEffectView.swift:

Canonical link: https://commits.webkit.org/293521@main
